### PR TITLE
add CartLineItems Subtotal

### PR DESCRIPTION
**Pull request by richardmonette**
Fixes https://github.com/Shopify/unity-buy-sdk/issues/148

Adds functionality for getting the cart subtotal. 

@Shopify/gaming

### CODE REVIEW

File: scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb [hunk 1]
```diff
@@ -147,7 +155,7 @@ namespace <%= namespace %>.SDK {
         ///
         /// cart.LineItems.AddOrUpdate(variantId, 3);
         /// \endcode
-        public void AddOrUpdate(string variantId, long? quantity = null, Dictionary<string, string> customAttributes = null) {
+        public void AddOrUpdate(string variantId, decimal price, long? quantity = null, Dictionary<string, string> customAttributes = null) {
             CartLineItem input = Get(variantId);
 
             if (input != null) {
```
Comment (dengjeffrey) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: Under this case when the cart line item already exists, should price be updated as well?

File: scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb [hunk 2]
```diff
@@ -213,13 +222,24 @@ namespace <%= namespace %>.SDK {
         /// cart.LineItems.AddOrUpdate(product, selectedOptions, 3);
         /// \endcode
         public void AddOrUpdate(Product product, Dictionary<string, string> selectedOptions, long? quantity = null, Dictionary<string, string> customAttributes = null) {
-            string variantId = VariantIdFromSelectedOptions(product, selectedOptions);
+            var variantId = VariantIdFromSelectedOptions(product, selectedOptions);
+            var variant = GetVariantById(product, variantId);
```
Comment (dengjeffrey) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: Would it be possible to refactor this so we get `ProductVariant` from one go? Versus making `VariantIdFromSelectedOptions` and `GetVariantById` which is two for loops over `Products.variants`

File: scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb [hunk 3]
```diff
@@ -48,12 +54,14 @@ namespace <%= namespace %>.SDK {
         public string ID = null;
         private string _VariantId;
         private long _Quantity;
+        private decimal _Price;
         private ObservableDictionary<string, string> _CustomAttributes;
         private OnCartLineItemChange OnChange;
-        
-        public CartLineItem(string variantId, OnCartLineItemChange onChange, long quantity = 1, IDictionary<string, string> customAttributes = null) {
+
+        public CartLineItem(string variantId, OnCartLineItemChange onChange, decimal price, long quantity = 1, IDictionary<string, string> customAttributes = null) {
```
Comment (dengjeffrey) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: This feels kind of shaky since we give power to the user to enter their own price, which is always a bad idea. What are the implications if instead `CartLineItem` takes in a `ProductVariant` instead of `variantID` and `price`. That way `CartLineItem` can get the variant ID, price, and any other variant info down the road

Comment (dengjeffrey) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: Same thing goes with removing `AddOrUpdate(string variantId, decimal price ...)` since we have `AddOrUpdate(ProductVariant variant, ...`

Comment (richardmonette) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: > This feels kind of shaky since we give power to the user to enter their own price, which is always a bad idea.

Definitely agree with you on this point.

So we've got three overloads of AddOrUpdate:

```public void AddOrUpdate(string variantId, long? quantity = null, Dictionary<string, string> customAttributes = null)

public void AddOrUpdate(ProductVariant variant, long? quantity = null, Dictionary<string, string> customAttributes = null)

public void AddOrUpdate(Product product, Dictionary<string, string> selectedOptions, long? quantity = null, Dictionary<string, string> customAttributes = null)
```

The later two are defined as basically calling through to the first version, where we actually use the `variantId`. I checked through our examples and, in all the cases I saw, we are recommending that developers use the second version. This makes the most sense too, imho, since when developers query products they get back a list of variants, so its reasonable they then takes those ProductVariant instances, and pass them back to Cart. 

With that in mind, I think we have two refactors here:

1. We make `public void AddOrUpdate(string variantId, long? quantity = null, Dictionary<string, string> customAttributes = null)` private (or remove it entirely). 
2. We change `CartLineItem` to accept a `ProductVariant`, instead of just an id.

@sleroux @dengjeffrey If you guys are good with that, I can do that as a separate, precursor PR to this.

File: scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb (hunk 1 above)
Comment (richardmonette) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: I think once we fix the main issue, by working on the ProductVariant level, instead of separate id and price, this sort of becomes less of an issue, but yes, I think it would update it.

File: scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb (hunk 2 above)
Comment (richardmonette) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: For sure, I think I'd break that out into another PR as well, as a precursor, so that can be a straight refactor which shouldn't be affecting any tests.

File: scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb (hunk 3 above)
Comment (dengjeffrey) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: Yes this sounds good, I personally prefer we remove it entirely. And use 
`public void AddOrUpdate(ProductVariant variant, long? quantity = null, Dictionary<string, string> customAttributes = null) ` as the base.

To me a `CartLineItem` is a `ProductVariant` with a quantity. Extracting properties of `ProductVariant` to create a `CartLineItem` causes confusion and adds room for error.

File: scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb [hunk 4]
```diff
@@ -391,14 +412,26 @@ namespace <%= namespace %>.SDK {
...
                 }
             }
 
             _IsSaved = true;
         }
 
+        private ProductVariant GetVariantById(Product product, string variantId) {
+            List<ProductVariant> variants = (List<ProductVariant>) product.variants();
+
+            foreach(var variant in variants) {
```
Comment (sleroux) [inline scripts/generator/graphql_generator/csharp/SDK/CartLineItems.cs.erb]: Nit: I think you can use `Find` here instead of iterating over the list yourself:

```csharp
var variants = (List<ProductVariant>)product.variants();
return variants.Find(variant => variant == variantID);
```
